pkg/importer/dataStream_test.go

```diff
@@ -7,416 +7,409 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
-	"log"
 	"math/rand"
 	"net/http"
+	"net/http/httptest"
 	"net/url"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"reflect"
+	"strconv"
 	"strings"
-	"testing"
-	"time"
 
+	"syscall"
+
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	"kubevirt.io/containerized-data-importer/pkg/common"
 	"kubevirt.io/containerized-data-importer/pkg/image"
+	"kubevirt.io/containerized-data-importer/pkg/util"
 	"kubevirt.io/containerized-data-importer/tests/utils"
 )
 
 const TestImagesDir = "../../tests/images"
 
+var imageDir, _ = filepath.Abs(TestImagesDir)
+var cirrosData, _ = readFile(filepath.Join(imageDir, "cirros-qcow2.img"))
+var tinyCoreData, _ = readFile(filepath.Join(imageDir, "tinyCore.iso"))
+var testfiles = createTestData()
+var stringRdr = strings.NewReader("test data for reader 1")
+var rdrfile, _ = os.Open(filepath.Join(imageDir, "cirros-qcow2.img"))
+var testFileRdrs = bufio.NewReader(rdrfile)
+
 type fakeQEMUOperations struct {
 	e1 error
 	e2 error
 	e3 error
 }
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRaw(string, string) error {
-	return o.e1
-}
+var _ = Describe("Data Stream", func() {
+	var ts *httptest.Server
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRawStream(*url.URL, string) error {
-	return o.e2
-}
+	BeforeEach(func() {
+		By("[BeforeEach] Creating test server")
+		ts = createTestServer(imageDir)
+	})
 
-func (o *fakeQEMUOperations) Validate(string, string) error {
-	return o.e3
-}
+	AfterEach(func() {
+		By("[AfterEach] closing test server")
+		ts.Close()
+	})
 
-func NewFakeQEMUOperations(e1, e2, e3 error) image.QEMUOperations {
-	return &fakeQEMUOperations{e1, e2, e3}
-}
+	table.DescribeTable("with import source should", func(image, accessKeyID, secretKey string, qemu bool, want []byte, wantErr bool) {
+		if image != "" {
+			image = ts.URL + "/" + image
+		}
+		By("Creating datastream")
+		ds, err := NewDataStream(image, accessKeyID, secretKey)
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			resultBuffer := make([]byte, len(want))
+			By("Reading data from buffer")
+			ds.Read(resultBuffer)
+			Expect(ds.accessKeyID).To(Equal(accessKeyID))
+			Expect(ds.secretKey).To(Equal(secretKey))
+			Expect(ds.qemu).To(Equal(qemu))
+			Expect(reflect.DeepEqual(resultBuffer, want)).To(BeTrue())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
```

Comment:
where are the Readers closed?

Reply:
Good point, they aren't being closed and should, let me add that.

Reply:
The Readers are now always closed in the AfterEach block